### ISSUE TITLE
Assert at ICSharpCode.Decompiler.IL.ILInstruction.CheckInvariant

### ISSUE THREAD

**Issue by greenozon** [icsharpcode/ILSpy#1616]
ILSpy version 5.0.0.5027-preview3

```
Error decompiling @06000674 ---.StringDecrypterInfo.Initialize
in assembly "C:\---"
 ---> ICSharpCode.ILSpy.AssertionFailedException: 
   at ICSharpCode.ILSpy.ILSpyTraceListener.Fail(String message, String detailMessage) in offset 380
   at ICSharpCode.ILSpy.ILSpyTraceListener.Fail(String message) in offset 9
   at System.Diagnostics.TraceInternal.Fail(String message) in offset 55
   at System.Diagnostics.Debug.Assert(Boolean condition) in offset 6
   at ICSharpCode.Decompiler.IL.ILInstruction.CheckInvariant(ILPhase phase) in offset 42
   at ICSharpCode.Decompiler.IL.Block.CheckInvariant(ILPhase phase) in offset 8
   at ICSharpCode.Decompiler.IL.ILInstruction.CheckInvariant(ILPhase phase) in offset 126
   at ICSharpCode.Decompiler.IL.IfInstruction.CheckInvariant(ILPhase phase) in offset 8
   at ICSharpCode.Decompiler.IL.ILInstruction.CheckInvariant(ILPhase phase) in offset 126
   at ICSharpCode.Decompiler.IL.Block.CheckInvariant(ILPhase phase) in offset 8
   at ICSharpCode.Decompiler.IL.Block.RunTransforms(IEnumerable`1 transforms, BlockTransformContext context) in offset 77
   at ICSharpCode.Decompiler.IL.Transforms.BlockILTransform.VisitBlock(ControlFlowNode cfgNode, BlockTransformContext context) in offset 146
   at ICSharpCode.Decompiler.IL.Transforms.BlockILTransform.VisitBlock(ControlFlowNode cfgNode, BlockTransformContext context) in offset 89
   at ICSharpCode.Decompiler.IL.Transforms.BlockILTransform.VisitBlock(ControlFlowNode cfgNode, BlockTransformContext context) in offset 89
   at ICSharpCode.Decompiler.IL.Transforms.BlockILTransform.VisitBlock(ControlFlowNode cfgNode, BlockTransformContext context) in offset 89
   at ICSharpCode.Decompiler.IL.Transforms.BlockILTransform.VisitBlock(ControlFlowNode cfgNode, BlockTransformContext context) in offset 89
   at ICSharpCode.Decompiler.IL.Transforms.BlockILTransform.Run(ILFunction function, ILTransformContext context) in offset 146
   at ICSharpCode.Decompiler.CSharp.CSharpDecompiler.DecompileBody(IMethod method, EntityDeclaration entityDecl, DecompileRun decompileRun, ITypeResolveContext decompilationContext) in offset 596
-- continuing with outer exception (ICSharpCode.Decompiler.DecompilerException) --
   at ICSharpCode.Decompiler.CSharp.CSharpDecompiler.DecompileBody(IMethod method, EntityDeclaration entityDecl, DecompileRun decompileRun, ITypeResolveContext decompilationContext) in offset 1084
   at ICSharpCode.Decompiler.CSharp.CSharpDecompiler.DoDecompile(IMethod method, DecompileRun decompileRun, ITypeResolveContext decompilationContext) in offset 249
   at ICSharpCode.Decompiler.CSharp.CSharpDecompiler.DoDecompile(ITypeDefinition typeDef, DecompileRun decompileRun, ITypeResolveContext decompilationContext) in offset 782
   at ICSharpCode.Decompiler.CSharp.CSharpDecompiler.DoDecompileTypes(IEnumerable`1 types, DecompileRun decompileRun, ITypeResolveContext decompilationContext, SyntaxTree syntaxTree) in offset 219
   at ICSharpCode.Decompiler.CSharp.CSharpDecompiler.DecompileTypes(IEnumerable`1 types) in offset 201
   at ICSharpCode.Decompiler.CSharp.WholeProjectDecompiler.<>c__DisplayClass30_0.<WriteCodeFilesInProject>b__2(IGrouping`2 file) in offset 69
   at System.Threading.Tasks.Parallel.<>c__DisplayClass17_0`1.<ForWorker>b__1() in offset 589
   at System.Threading.Tasks.Task.InnerInvokeWithArg(Task childTask) in offset 6
   at System.Threading.Tasks.Task.<>c__DisplayClass176_0.<ExecuteSelfReplicating>b__0(Object <p0>) in offset 134

```

https://www.sendspace.com/file/o11ji1

**Comment by dgrunwald** [comment]
Failing assert is `Debug.Assert(child.Parent == this);`
```
this = {Block IL_0142 {
	throw(newobj ApplicationException..ctor(ldstr "Could not find string offset"))
} at IL_0142}

child.Parent = {if.notnull(call FindOffsetValue(ldloc this, ldloc V_0), throw(newobj ApplicationException..ctor(ldstr "Could not find string offset")))}
```
So it's related to the recently merged throw-expression support.

**Comment by siegfriedpammer** [comment]
I fixed a similar bug in 3ea2ce4e2d28884310222752f295a382b9fc16f1

**Comment by dgrunwald** [comment]
Yeah, looks like this is the same, but for the `InstructionCollection` indexer setter.
